File: plugins/csp-visual-query/src/sourceNodes/WCSImageLoader.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
//                               This file is part of CosmoScout VR                               //
////////////////////////////////////////////////////////////////////////////////////////////////////

// SPDX-FileCopyrightText: German Aerospace Center (DLR) <cosmoscout@dlr.de>
// SPDX-License-Identifier: MIT

#include "WCSImageLoader.hpp"

#include "../../../../src/cs-utils/filesystem.hpp"
#include "../../../csl-ogc/src/wcs/WebCoverageService.hpp"
#include "../../../csl-ogc/src/wcs/WebCoverageTextureLoader.hpp"
#include "../types/types.hpp"

namespace csp::visualquery {

////////////////////////////////////////////////////////////////////////////////////////////////////

const std::string WCSImageLoader::sName = "WCSImageLoader";

////////////////////////////////////////////////////////////////////////////////////////////////////

std::string WCSImageLoader::sSource() {
  return cs::utils::filesystem::loadToString(
      "../share/resources/nodes/csp-visual-query/WCSImageLoader.js");
}

////////////////////////////////////////////////////////////////////////////////////////////////////

std::unique_ptr<WCSImageLoader> WCSImageLoader::sCreate(
  std::shared_ptr<std::vector<csl::ogc::WebCoverageService>> wcs) {
  return std::make_unique<WCSImageLoader>(std::move(wcs));
}

////////////////////////////////////////////////////////////////////////////////////////////////////

WCSImageLoader::WCSImageLoader(std::shared_ptr<std::vector<csl::ogc::WebCoverageService>> wcs)
  : mWcs(std::move(wcs))
  , mSelectedServer(nullptr)
  , mSelectedImageChannel(nullptr) {
}

////////////////////////////////////////////////////////////////////////////////////////////////////

WCSImageLoader::~WCSImageLoader() {
}

////////////////////////////////////////////////////////////////////////////////////////////////////

std::string const& WCSImageLoader::getName() const {
  return sName;
}

////////////////////////////////////////////////////////////////////////////////////////////////////

void WCSImageLoader::onMessageFromJS(nlohmann::json const& message) {

  logger().debug("Message form JS: {}", message.dump());

  // Send available servers
  if (message.dump() == "\"requestServers\"") {
    sendServersToJs();
    return;
  }

  // set newly selected server
  if (message.find("server") != message.end()) {

    // reset server and image channel selection    
    if (message["server"] == "none") {
      mSelectedServer = nullptr;
      mSelectedImageChannel = nullptr;
      
      nlohmann::json imageChannel;
      imageChannel["imageChannel"] = "reset";
      sendMessageToJS(imageChannel);
    
    // set new server and send available image channels
    } else {
      for (auto wcs : (*mWcs)) {
        if (wcs.getTitle() == message["server"]) {
          mSelectedServer = std::make_shared<csl::ogc::WebCoverageService>(wcs);
          break;
        }
      }
      sendImageChannelsToJs();
    }
    return;
  }

  // set newly selected image channel
  else if (message.find("imageChannel") !=  message.end()) {
    
    // reset image channel selection
    if (message["imageChannel"] == "none") {
      mSelectedImageChannel = nullptr;

    // set new image channel
    } else {
      auto temp = mSelectedServer->getCoverage(message["imageChannel"]);
      mSelectedImageChannel = (temp.has_value() ? std::make_shared<csl::ogc::WebCoverage>(temp.value()) : nullptr);
    }
  }

  // Whenever the server changes, we write the new output by calling the process() method.
  // Writing the output will not trigger a graph reprocessing right away, it will only queue
  // up the connected nodes for being processed in the next update step.
  process();
}

////////////////////////////////////////////////////////////////////////////////////////////////////

nlohmann::json WCSImageLoader::getData() const {
  nlohmann::json data;
  if (mSelectedServer != nullptr) {
    data["serverUrl"] = mSelectedServer->getUrl();

    if (mSelectedImageChannel != nullptr) {
      data["imageChannelId"] = mSelectedImageChannel->getId();
    }
  }
  return data;
}

////////////////////////////////////////////////////////////////////////////////////////////////////

void WCSImageLoader::setData(nlohmann::json const& json) {
  if (json.find("serverUrl") != json.end()) {

    for (csl::ogc::WebCoverageService wcs : *mWcs) {
      if (wcs.getUrl() == json["serverUrl"]) {
        mSelectedServer = std::make_shared<csl::ogc::WebCoverageService>(wcs);
        break;
      }
    }

    if (mSelectedServer != nullptr && json.find("imageChannelId") != json.end()) {

      auto temp = mSelectedServer->getCoverage(json["imageChannelId"]);
      mSelectedImageChannel = (temp.has_value() ? std::make_shared<csl::ogc::WebCoverage>(temp.value()) : nullptr);
    }
  }
  // process() ?
}

////////////////////////////////////////////////////////////////////////////////////////////////////

void WCSImageLoader::process() {
  // double first  = readInput<double>("first", 0.0);

  // The name of the port must match the name given in the JavaScript code above.
  // writeOutput("minDataValue", 0);

  // create request for texture loading
  if (mSelectedServer == nullptr || mSelectedImageChannel == nullptr) {
    return;
  }

  logger().debug("process!");

  csl::ogc::WebCoverageTextureLoader::Request request;
  
  request.mTime = std::to_string(readInput<double>("wcsTime", 0.0));
  
  logger().debug(0);

  csl::ogc::Bounds bound;
  bound.mMinLon = readInput<double>("xBoundMin", 0.0);
  bound.mMaxLon = readInput<double>("xBoundMax", 100.0);
  bound.mMinLat = readInput<double>("yBoundMin", 0.0);
  bound.mMaxLat = readInput<double>("yBoundMax", 100.0);
  request.mBounds = bound;

  logger().debug(1);

  request.mMaxSize = readInput<int>("resolution", 1024);

  logger().debug(2);

  request.mFormat = "image/tiff";

  logger().debug(3);

  auto texLoader = csl::ogc::WebCoverageTextureLoader();
  auto texture = texLoader.loadTexture(*mSelectedServer, *mSelectedImageChannel, request, 
    "../../../install/windows-Release/share/cache/csp-visual-query/texture-cache", true);

  logger().debug(4);

  if (texture.has_value()) {
    logger().debug("x in texture: {}", texture.value().x);
    logger().debug("y in texture: {}", texture.value().y);
  }

  logger().debug(5);
}

void WCSImageLoader::sendServersToJs() {
  std::vector<std::string> serverNames;
  for (csl::ogc::WebCoverageService server : *mWcs) {
    serverNames.push_back(server.getTitle());
  }
  nlohmann::json server;
  server["server"] = serverNames;
  sendMessageToJS(server);
}

////////////////////////////////////////////////////////////////////////////////////////////////////

void WCSImageLoader::sendImageChannelsToJs() {
  std::vector<std::string> imageChannelNames;
  
  for (csl::ogc::WebCoverage imageChannel : mSelectedServer->getCoverages()) {
    imageChannelNames.push_back(imageChannel.getTitle());
  }
  nlohmann::json imageChannels;
  imageChannels["imageChannel"] = imageChannelNames;
  sendMessageToJS(imageChannels);
}

} // namespace csp::visualquery

File: plugins/csp-visual-query/src/sourceNodes/WCSImageLoader.js
////////////////////////////////////////////////////////////////////////////////////////////////////
//                               This file is part of CosmoScout VR                               //
////////////////////////////////////////////////////////////////////////////////////////////////////

// SPDX-FileCopyrightText: German Aerospace Center (DLR) <cosmoscout@dlr.de>
// SPDX-License-Identifier: MIT

// The TimeNode is pretty simple as it only has a single output socket. The component serves as
// a kind of factory. Whenever a new node is created, the builder() method is called.
class WCSImageLoaderComponent extends Rete.Component {

  constructor() {
    // This name must match the WCSSourceNode::sName defined in WCSSource.cpp.
    super("WCSImageLoader");

    // This specifies the submenu from which this node can be created in the node editor.
    this.category = "Sources";
  }

  // Called whenever a new node of this type needs to be constructed.
  builder(node) {

    // This node has a single output. The first parameter is the name of this output and must be
    // unique amongst all sockets. It is also used in the WCSImageLoader::process() to write the
    // output of this node. The second parameter is shown as name on the node. The last
    // parameter references a socket type which has been registered with the node factory
    // before. It is required that the class is called <NAME>Component.

    let xBoundMinInput = new Rete.Input('xBoundMin', "Longitude Min", CosmoScout.socketTypes['number Value']);
    node.addInput(xBoundMinInput);

    let xBoundMaxInput = new Rete.Input('xBoundMax', "Longitude Max", CosmoScout.socketTypes['number Value']);
    node.addInput(xBoundMaxInput);

    let yBoundMinInput = new Rete.Input('yBoundMin', "Latitude Min", CosmoScout.socketTypes['number Value']);
    node.addInput(yBoundMinInput);

    let yBoundMaxInput = new Rete.Input('yBoundMax', "Latitude Max", CosmoScout.socketTypes['number Value']);
    node.addInput(yBoundMaxInput);
    
    let timeInput = new Rete.Input('wcsTime', "Time", CosmoScout.socketTypes['WCSTime']);
    node.addInput(timeInput);
    
    let resolutionInput = new Rete.Input('resolution', "Resolution", CosmoScout.socketTypes['WCSResolution']);
    node.addInput(resolutionInput);

    let imageOutput = new Rete.Output('image', 'Image 2D', CosmoScout.socketTypes['Image2D']);
    node.addOutput(imageOutput);
    
    let minOutput = new Rete.Output('minDataValue', 'Min', CosmoScout.socketTypes['number Value']);
    node.addOutput(minOutput);

    let maxOutput = new Rete.Output('maxDataValue', 'Max', CosmoScout.socketTypes['number Value']);
    node.addOutput(maxOutput);

    let serverControl = new ServerControl('selectServer');
    node.addControl(serverControl);

    let imageChannelControl = new ImageChannelControl('selectImageChannel');
    node.addControl(imageChannelControl);

    node.onInit = (nodeDiv) => { 

      serverControl.init(nodeDiv, node.data); 
      imageChannelControl.init(nodeDiv, node.data);

      node.onMessageFromCPP = (message) => {

        // display servers in dropdown
        if (message["server"]) {
          node.data.url = message["server"];
          serverControl.createServerSelection(message["server"]);  
        }

        // new image channels received
        else if (message["imageChannel"]) {

          // rest image channels selection
          if (message["imageChannel"] === "reset") {
            node.data.imageChannel = [];
            imageChannelControl.resetImageChannelSelection();  

          // add new image channels to dropdown
          } else {
            node.data.imageChannel = message["imageChannel"];        
            imageChannelControl.createImageChannelSelection(message["imageChannel"]);
          }
        
        } else {
          console.log("Unknown cpp message:");
          console.log(message);
        }
      };
    }

    let button1 = document.createElement("button");
    button1.innerHTML = "request server";
    button1.addEventListener("click", (e) => {
      serverControl.sendServerRequest();
    });
    document.body.insertBefore(button1, document.body.firstChild);

    return node;
  }
}

// This is the widget which is used for selecting the server.
class ServerControl extends Rete.Control {
  constructor(key) {
    super(key);
    this.selectElement;

    // This HTML code will be used whenever a node is created with this widget.
    this.template = `
          <select>
            <option value="none">None</option>
          </select>

          <style>
            .dropdown {
              margin: 10px 15px !important;
              width: 150px !important;
            }
          </style>
        `;
  }

  // This is called by the node.onInit() above once the HTML element for the node has been
  // created. If present, the data object may contain a math operation as returned by
  // MathNode::getData() which - if present - should be preselected.
  init(nodeDiv, data) {

    // Initialize the bootstrap select.
    this.selectElement = nodeDiv.querySelector("select");
    $(this.selectElement).selectpicker();

    // Preselect a server.
    if (data.url) {
      $(this.selectElement).selectpicker('val', data.url);
    }

    // Send an update to the node editor server whenever the user selects a new server.
    this.selectElement.addEventListener('change',
      (e) => {
        CosmoScout.sendMessageToCPP({server: e.target.value}, this.parent.id); 
      });
  }

  // Send a request to get the available servers.
  sendServerRequest() {
    if (!this.parent.data.url) {
      CosmoScout.sendMessageToCPP("requestServers", this.parent.id); 
    }
  }

  createServerSelection(values) {
    // add new elements
    for (let i = 0; i < values.length; i++) {
      let option = document.createElement("option");
      option.value = values[i]; 
      option.innerHTML = values[i];
      this.selectElement.appendChild(option);
    }
    // refresh bootstrap dropdown options
    $(this.selectElement).selectpicker("refresh");     
  }
}

// This is the widget which is used for selecting the image Channel.
class ImageChannelControl extends Rete.Control {
  constructor(key) {
    super(key);
    this.selectElement;

    // This HTML code will be used whenever a node is created with this widget.
    this.template = `
          <select>
            <option value="0">None</option>
          </select>

          <style>
            .dropdown {
              margin: 10px 15px !important;
              width: 150px !important;
            }
          </style>
        `;
  }

  // This is called by the node.onInit() above once the HTML element for the node has been
  // created. If present, the data object may contain a math operation as returned by
  // MathNode::getData() which - if present - should be preselected.
  init(nodeDiv, data) {

    // Initialize the bootstrap select.
    this.selectElement = nodeDiv.querySelectorAll("select")[1];

    $(this.selectElement).selectpicker();

    // Preselect a server.
    if (data.imageChannel) {
      $(this.selectElement).selectpicker('val', data.imageChannel);
    }

    // Send an update to the node editor server whenever the user selects a new channel.
    this.selectElement.addEventListener('change',
      (e) => {
        CosmoScout.sendMessageToCPP({imageChannel: e.target.value}, this.parent.id); 
      });
  }

  createImageChannelSelection(values) {    
    this.resetImageChannelSelection();

    // add new elements
    for (let i = 0; i < values.length; i++) {
      let option = document.createElement("option");
      option.value = values[i]; 
      option.innerHTML = values[i];
      this.selectElement.appendChild(option);
    }
    // refresh bootstrap dropdown options
    $(this.selectElement).selectpicker("refresh");
  }

  resetImageChannelSelection() {
    // remove all elements
    while (this.selectElement.firstChild) {
      this.selectElement.removeChild(this.selectElement.lastChild);
    }

    // add "None" element
    let none = document.createElement("option");
    none.value = 0; 
    none.innerHTML = "None";
    this.selectElement.appendChild(none);

    // refresh bootstrap dropdown options
    $(this.selectElement).selectpicker("refresh");
  }
}